apps/asap-server/test/handlers/research-output/index-handler.test.ts

```diff
@@ -20,36 +20,127 @@ describe('Research Output index handler', () => {
     const researchOutputResponse = getResearchOutputResponse();
     researchOutputControllerMock.fetchById.mockResolvedValueOnce(
       researchOutputResponse,
     );
 
-    await indexHandler(getEvent());
+    await indexHandler(
+      createEventBridgeEventMock(
+        {
+          type: 'ResearchOutputsPublished',
+          payload: {
+            $type: 'EnrichedContentEvent',
+            type: 'Published',
+            id: '0ecccf93-bd06-4307-90ea-c153fe495580',
+          },
+        },
+        'ResearchOutputCreated',
+      ),
+    );
 
     expect(algoliaIndexMock.saveObject).toHaveBeenCalledWith({
       ...researchOutputResponse,
       objectID: researchOutputResponse.id,
     });
   });
-});
 
-const getEvent = (): EventBridgeEvent<
-  ResearchOutputEventType,
-  SquidexWebhookResearchOutputPayload
-> =>
-  createEventBridgeEventMock(
-    createResearchOutputSquidexWebhookPayload,
-    'ResearchOutputCreated',
-  );
+  test('Should fetch the research-output and create a record in Algolia when research-output is updated', async () => {
+    const researchOutputResponse = getResearchOutputResponse();
+    researchOutputControllerMock.fetchById.mockResolvedValueOnce(
+      researchOutputResponse,
+    );
+
+    await indexHandler(
+      createEventBridgeEventMock(
+        {
+          type: 'ResearchOutputsUpdated',
+          payload: {
+            $type: 'EnrichedContentEvent',
+            type: 'Published',
+            id: '0ecccf93-bd06-4307-90ea-c153fe495580',
+          },
+        },
+        'ResearchOutputUpdated',
+      ),
+    );
+
+    expect(algoliaIndexMock.saveObject).toHaveBeenCalledWith({
+      ...researchOutputResponse,
+      objectID: researchOutputResponse.id,
+    });
+  });
+
+  test('Should remove the record Algolia when research-output when research-output is not found', async () => {
+    researchOutputControllerMock.fetchById.mockRejectedValueOnce({
+      statusCode: 404,
+    });
+
+    const event: EventBridgeEvent<
+      ResearchOutputEventType,
+      SquidexWebhookResearchOutputPayload
+    > = createEventBridgeEventMock(
+      {
+        type: 'ResearchOutputsUnpublished',
+        payload: {
+          $type: 'EnrichedContentEvent',
+          type: 'Unpublished',
+          id: '0ecccf93-bd06-4307-90ea-c153fe495580',
+        },
+      },
+      'ResearchOutputDeleted',
+    );
+
+    await indexHandler(event);
 
-const createResearchOutputSquidexWebhookPayload: SquidexWebhookResearchOutputPayload =
-  {
-    type: 'ResearchOutputsCreated',
-    payload: {
-      $type: 'EnrichedContentEvent',
-      type: 'Created',
-      id: '0ecccf93-bd06-4307-90ea-c153fe495580',
-    },
-  };
+    expect(algoliaIndexMock.deleteObject).toHaveBeenCalledWith(
+      event.detail.payload.id,
+    );
+  });
+
+  test('Should not relay on the order of the events, i.e. unpublished and published', async () => {
```

Comment:
i think this test doesn't really assert what you want anyway - do you know when the delete and save methods were called? You would probably want to write the expectations after the first and second invocation.

Anyway I feel like we need to test the following scenarios:

1. User unpublishes and quickly publishes - the events come in the right order
In such scenario we will invoke the handler with the unpublished and published event. Because the user quickly published the entity back - the resource will always be present. We want to make sure that `saveObject` method was called with the last invocation (might be called twice) - we could also make sure that the `deleteObject` method doesn't get called since the entity is always present.

2. User unpublishes and quickly publishes - the events come in the reverse order 
The order of the invocations is reversed however we still want to make sure that the `saveObject` method was called last  (it might even get called twice like previously) and again - we could check that the `deleteObject` method never gets called.

3. User publishes and quickly unpublishes - the events come in the right order
In here we want to make sure that the entity is eventually deleted. When the handlers receive the notifications the entity will not be present - btw can you make sure that the `deleteObject` method doesnt throw when the resource isn't present (i doubt that it does)? Technically in this scenario we could also assert that the `saveObject` method never gets called

4. User publishes and quickly unpublishes - the events come in the reverse order
I think you know where I am going with this so...

Basically we want to make sure that we capture the race conditions - I think your current implementation where the object is saved or deleted irrespective of the received event (it only depends on the current object state) is actually correct but I didn't give it much thought - this will come up in tests anyway.

I will make sure to add info to the ticket that these scenarios need to be tested  - I would recommend that you manually test it too before sending across to AC.

Maybe you can think of other scenarios that we want to test? I generally prefer to have too many tests than too few